src/edu/washington/escience/myria/operator/Apply.java

```diff
@@ -153,14 +182,29 @@ protected TupleBatch fetchNextReady() throws DbException, InvocationTargetExcept
             if (!emptyProduct) {
               // Initialize each iterator to its starting index.
               Arrays.fill(iteratorIndexes, 0);
-              // Iterate over each element of the Cartesian product and append to output.
+              // Iterate over each element of the Cartesian product and append to output
+              TupleBuffer countIdx = null;
+              int iteratorIdx = 0;
+              int resultRowIdx = 0;
+              int flatmapid = -1;
               do {
-                for (int iteratorIdx = 0; iteratorIdx < iteratorIndexes.length; ++iteratorIdx) {
-                  int resultRowIdx =
-                      lastCumResultCounts[iteratorIdx] + iteratorIndexes[iteratorIdx];
+                for (iteratorIdx = 0; iteratorIdx < iteratorIndexes.length; ++iteratorIdx) {
+                  resultRowIdx = lastCumResultCounts[iteratorIdx] + iteratorIndexes[iteratorIdx];
+                  if (flatmapid < iteratorIndexes[iteratorIdx]) {
```

Comment:
not sure I understand  your example, this is  included in the tests and  example json for regular apply as well as flatmap apply.
basically if each input tuple returns a single tuple, the flatmap id is exactly the same for all tuples. if however an input tuple returns multiple tuples( n tuples), each of the return tuples have a flatmapid from 0 to n-1.